File: android/Sightread/app/src/main/java/com/meta/wearable/dat/externalsampleapps/sightread/ai/VisionAIService.kt
package com.meta.wearable.dat.externalsampleapps.sightread.ai

import android.graphics.Bitmap
import java.io.ByteArrayOutputStream

interface VisionAIService {
  suspend fun analyze(jpegData: ByteArray, prompt: String): String
}

object VisionAIServiceFactory {
  fun create(provider: AIProvider, settings: SettingsRepository): VisionAIService =
      when (provider) {
        AIProvider.GEMINI -> GeminiVisionService(settings.geminiApiKey)
        AIProvider.OPENAI -> OpenAIVisionService(settings.openAIApiKey)
        AIProvider.GROQ -> GroqVisionService(settings.groqApiKey)
      }
}

object ImageEncoding {
  fun jpegBytes(bitmap: Bitmap, maxWidth: Int = 512, quality: Int = 60): ByteArray? {
    val scale = minOf(1f, maxWidth.toFloat() / bitmap.width)
    val w = (bitmap.width * scale).toInt().coerceAtLeast(1)
    val h = (bitmap.height * scale).toInt().coerceAtLeast(1)
    val scaled = Bitmap.createScaledBitmap(bitmap, w, h, true)
    return ByteArrayOutputStream().use { stream ->
      if (!scaled.compress(Bitmap.CompressFormat.JPEG, quality, stream)) return null
      stream.toByteArray()
    }
  }
}

File: ios/Sightread/Sightread/Services/VisionAIService.swift
import Foundation
import UIKit

protocol VisionAIService: Sendable {
  func analyze(jpegData: Data, prompt: String) async throws -> String
}

enum VisionAIError: LocalizedError {
  case missingAPIKey, invalidResponse
  case httpError(Int, String)
  var errorDescription: String? {
    switch self {
    case .missingAPIKey: return "Add an API key in Settings."
    case .invalidResponse: return "Could not parse the AI response."
    case .httpError(let c, let b): return "API error \(c): \(b)"
    }
  }
}

enum VisionAIServiceFactory {
  static func make(provider: AIProvider, settings: SettingsStore) -> VisionAIService {
    switch provider {
    case .gemini: return GeminiVisionService(apiKey: settings.geminiAPIKey)
    case .openai: return OpenAIVisionService(apiKey: settings.openAIAPIKey)
    case .groq: return GroqVisionService(apiKey: settings.groqAPIKey)
    }
  }
}

enum ImageEncoding {
  static func jpegData(from image: UIImage, maxWidth: CGFloat = 512, quality: CGFloat = 0.6) -> Data? {
    let size = image.size
    guard size.width > 0, size.height > 0 else { return nil }
    let scale = min(1, maxWidth / size.width)
    let target = CGSize(width: size.width * scale, height: size.height * scale)
    let resized = UIGraphicsImageRenderer(size: target).image { _ in image.draw(in: CGRect(origin: .zero, size: target)) }
    return resized.jpegData(compressionQuality: quality)
  }
}